### PR TITLE
Allow caching options from environment variable

### LINKED ISSUE #2709: How enable cache from environment variable?
**Issue type:**

[x] question
[ ] bug report
[x] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[x] `0.2.7` (or put your version here)

**Question**

How set `cache: true` option from `.env` file? 
Maybe add support `TYPEORM_CACHE` environment variable?

### PR DESCRIPTION
**Pull request by rustamwin**
Fixes #2709

### CODE REVIEW

**Comment by vlapo** [comment]
Missing docs. And I think we should provide small description in docs about this options

File: src/connection/options-reader/ConnectionOptionsEnvReader.ts [hunk 1]
```diff
@@ -43,7 +43,13 @@ export class ConnectionOptionsEnvReader {
...
                 migrationsDir: PlatformTools.getEnvVariable("TYPEORM_MIGRATIONS_DIR"),
                 subscribersDir: PlatformTools.getEnvVariable("TYPEORM_SUBSCRIBERS_DIR"),
-            }
+            },
+            cache: (typeof this.transformCaching(PlatformTools.getEnvVariable("TYPEORM_CACHE")) === "string") ? {
+                type: this.transformCaching(PlatformTools.getEnvVariable("TYPEORM_CACHE")),
+                options: PlatformTools.getEnvVariable("TYPEORM_CACHE_OPTIONS") ? JSON.parse(PlatformTools.getEnvVariable("TYPEORM_CACHE_OPTIONS")) : undefined,
+                alwaysEnabled: this.transformCaching(PlatformTools.getEnvVariable("TYPEORM_CACHE_ALWAYS_ENABLED")),
+                duration: parseInt(PlatformTools.getEnvVariable("TYPEORM_CACHE_DURATION"))
+            } : this.transformCaching(PlatformTools.getEnvVariable("TYPEORM_CACHE"))
```
[the remark this answers is not in the document]
Comment (rustamwin) [reply src/connection/options-reader/ConnectionOptionsEnvReader.ts]: agree

File: src/connection/options-reader/ConnectionOptionsEnvReader.ts [hunk 2]
```diff
@@ -63,6 +69,20 @@ export class ConnectionOptionsEnvReader {
         return this.stringToArray(logging);
     }
 
+    /**
+     * Transforms logging string into real logging value connection requires.
+     */
+    protected transformCaching(caching: string): any {
```
[the remark this answers is not in the document]
Comment (rustamwin) [reply src/connection/options-reader/ConnectionOptionsEnvReader.ts]: No, this function returns bool or string or undefined

[the remark this answers is not in the document]
Comment (vlapo) [reply src/connection/options-reader/ConnectionOptionsEnvReader.ts]: So use `string | boolean | undefined` no `any`. 

We are using typescript so use types everytime it is possible. :) I know this is just simple code and it is affect nothing but in complex code it should be important. So lets be be a little bit strict and keep some code hygiene.

[the remark this answers is not in the document]
Comment (rustamwin) [reply src/connection/options-reader/ConnectionOptionsEnvReader.ts]: Ok. I fixed it

**Comment by rustamwin** [comment]
> Missing docs

Don't understand, typedoc?

**Comment by vlapo** [comment]
Update `docs/using-ormconfig.md`. Do not forget mention `TYPEORM_CACHE` is possible boolean or cache type string

**Comment by vlapo** [comment]
Why? Its few lines in docs, so do it in one PR and have everything in one place.

**Comment by rustamwin** [comment]
@vlapo Done

**Comment by vlapo** [comment]
Thnx @rustamwin. I know, I bother you with the little things. But I want to only high class PR and better code quality.

**Comment by drwatsno** [comment]
Using environment variables from the library is an antipattern. The library must be configured explicitly - through the settings. Environment variables must be read by your application and passed to TypeORM.

Read about 12factor apps https://12factor.net/
Take a look at https://github.com/mozilla/node-convict

Also, don't read any .env files in production code, this violates the principles of 12factor applications.

**Comment by pleerock** [comment]
Im going to merge this PR before this discussion proceed.

**Comment by pleerock** [comment]
@drwatsno we already had a discussion with @Kononnable recently on a current state of configuration in typeorm and decided to merge PRs that are related to configuration for now since all functionality is in place at the moment.

I don't like current state of configuration, and TBH I would really like to remove it from this library. The only thing I want to have is just `createConnection` method that accepts configuration object. That't it. But there are two questions:

* user convenience 
* and what to do with CLI which is heavily rely on `ormconfig` to execute most of commands. Passing all configuration to cli each time obviously is a bad idea

So I would really like to do what I wrote above and at the same time resolve those two issues.

**Comment by drwatsno** [comment]
> @drwatsno well, your argument makes some sense, but why can’t typeorm use env variables? Many libraries do this, it simplifies the work and 12factors claim to use env variables

It is true that many libraries use this approach. I will try to explain my opinion more clearly. For example, you have library A, which defines the variables ENV1 and ENV2. And another library - B, in which there is ENV3. This is normal as long as there are no collisions in the names, but history shows that sometimes variable names overlap. That's why people started doing something like A_ENV1 and B_ENV3.

Another example is where you have some env in CI that can intersect with a particular library (for example, in Gitlab you can have project level variables for library A, and then some variables defined in the repository). At some point, you include library A in your application (and you don’t even know about project-level variables), and it begins to implicitly use project-level variables.

Such implicit behavior can lead to various errors that are very difficult to debug.

As a rule, you must configure the application through one point. This is understandable, easy to use and explain to new people in your project. Imagine the situation - Bob begins his work on your application, and you explain to him:

- Hi, Bob, to look at ORM envs, please look there, and you can also adjust the logging level for this library, please see the logs library documentation. And, of course, take a look at the rpc library, because we have defined some environment variables for the discovery strategy, etc.

The point is not even in 12 factors, but in the principles of software design.
P.S. At some point, you will update one of your dependencies, which defines such variables, and find that one of them is changed. Of course, you will know about it when your application fail in production.

**Comment by havenchyk** [comment]
thanks for clarification @drwatsno. it totally makes sense, I still don't see the good solution with passing env variables to cli commands, but as for general idea to use env variables to fill values in ormconfig I absolutely agree

**Pull request merged**